onnx/ir.h

```diff
@@ -0,0 +1,1048 @@
+    return inputs_.at(0);
+  }
+  Value * output() {
+    ONNX_ASSERT(outputs_.size() == 1);
+    return outputs_.at(0);
+  }
+  const  Value * input() const {
+    ONNX_ASSERT(inputs_.size() == 1);
+    return inputs_.at(0);
+  }
+  // Access a particular input.  This is a checked index.
+  Value * input(size_t i) {
+    return inputs_.at(i);
+  }
+  const Value * input(size_t i) const {
+    return inputs_.at(i);
+  }
+
+  // Graphs
+
+  // Note [Topological invariant]
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // We always maintain an up-to-date topological ordering of all nodes via
+  // the next()/prev() links.  All transformations to graphs must preserve
+  // this topological ordering: for example, it is only valid to 'addInput'
+  // with an input which is topologically before the current node.
+  //
+  // Usually, it is obvious whether or not topological order is maintained;
+  // for example, if you are adding nodes to the end of the topsort, it's
+  // impossible for them to refer to inputs that are not in the topsort.
+  // If it is not obvious, please comment accordingly.
+
+  // Add 'node' as an input to 'this' at the end of existing
+  // arguments.  Returns the added node for ease of chaining.
+  //
+  // Given:   %3 = f(%1, %2)
+  // Execute: %3.addInput(%4)
+  // Result:  %3 = f(%1, %2, %4)
+  Value* addInput(Value * node) {
+    ONNX_ASSERT(graph_ == node->owningGraph());
+    node->uses_.emplace_back(this, inputs_.size());
+    inputs_.push_back(node);
+    return node;
+  }
+
+  // Replace the input of 'this' at position 'i' with
+  // 'newValue', returning the old node.
+  //
+  // Given:   %3 = f(%1, %2)
+  // Execute: %3.replaceInput(1, %4)
+  // Result:  %3 = f(%1, %4)
+  Value * replaceInput(size_t i, Value * newValue) {
+    ONNX_ASSERT(newValue->owningGraph() == graph_);
+    Value * old = dropInput(i);
+    inputs_[i] = newValue;
+    newValue->uses_.emplace_back(this, i);
+    return old;
+  }
+
+  // Replace all occurrences of 'from' in the inputs of this
+  // node with 'to'. Corresponds to llvm's replaceUsesOfWith.
+  //
+  // Given:   %3 = f(%1, %2, %1)
+  // Execute: %3.replaceInputWith(%1, %4)
+  // Result:  %3 = f(%4, %2, %4)
+  void replaceInputWith(Value * from, Value * to) {
+    ONNX_ASSERT(from->owningGraph() == graph_);
+    ONNX_ASSERT(to->owningGraph() == graph_);
+    size_t i = 0;
+    for(auto input : inputs()) {
+      if(input == from)
+        replaceInput(i, to);
+      i++;
+    }
+  }
+
+  Value* addOutput() {
+    outputs_.push_back(new Value(this, outputs_.size()));
+    return outputs_.back();
+  }
+
+  void eraseOutput(size_t i);
+
+  // Insert unattached 'this' node after 'n' in the topological order.
+  // Returns this (for chaining).
+  //
+  // Given:   %3 = f(%1, %2)
+  //          %4 = g(%3)
+  // and unattached: %5 = h(%1)
+  // Execute: %5.insertBefore(%4)
+  // Result:  %3 = f(%1, %2)
+  //          %5 = h(%1)
+  //          %4 = g(%3)
+  Node* insertBefore(Node * n) {
```

Comment:
these insert/move APIs may be APIs in graph level. Move it to Graph?

Reply:
Every Node is uniquely owned by a Graph, so it's less typing to do it this way.